packages/ember-template-compiler/lib/plugins/deprecate-with.ts

```diff
@@ -0,0 +1,189 @@
+import { assert, deprecate } from '@ember/debug';
+import { AST, ASTPlugin } from '@glimmer/syntax';
+import calculateLocationDisplay from '../system/calculate-location-display';
+import { EmberASTPluginEnvironment } from '../types';
+import { isPath } from './utils';
+
+/**
+ @module ember
+*/
+
+/**
+  A Glimmer2 AST transformation that deprecates `{{#with}}` and replace it
+  with `{{#let}}` and `{{#if}}` as per RFC 445.
+
+  Transforms:
+
+  ```handlebars
+  {{#with this.foo as |bar|}}
+    ...
+  {{/with}}
+  ```
+
+  Into:
+
+  ```handlebars
+  {{#let this.foo as |bar|}}
+    {{#if bar}}
+      ...
+    {{/if}}
+  {{/let}}
+  ```
+
+  And:
+
+  ```handlebars
+  {{#with this.foo as |bar|}}
+    ...
+  {{else}}
+    ...
+  {{/with}}
+  ```
+
+  Into:
+
+  ```handlebars
+  {{#let this.foo as |bar|}}
+    {{#if bar}}
+      ...
+    {{else}}
+      ...
+    {{/if}}
+  {{/let}}
+  ```
+
+  And issues a deprecation message.
+
+  @private
+  @class DeprecateWith
+*/
+export default function deprecateWith(env: EmberASTPluginEnvironment): ASTPlugin {
+  let { moduleName } = env.meta;
+  let { builders: b } = env.syntax;
+
+  return {
+    name: 'deprecate-with',
+
+    visitor: {
+      BlockStatement(node: AST.BlockStatement) {
+        if (!isPath(node.path) || node.path.original !== 'with') return;
+
+        let { params, hash, program, inverse, loc, openStrip, inverseStrip, closeStrip } = node;
+
+        let sourceInformation = calculateLocationDisplay(moduleName, node.loc);
+
+        assert(
+          `\`{{#with}}\` takes a single positional argument (the path to alias), received ${displayParams(
+            params
+          )}. ${sourceInformation}`,
+          params.length === 1
+        );
+
+        assert(
+          `\`{{#with}}\` does not take any named arguments, received ${displayHash(
+            hash
+          )}. ${sourceInformation}`,
+          hash.pairs.length === 0
+        );
+
+        assert(
+          `\`{{#with}}\` yields a single block param, received ${displayBlockParams(
+            program.blockParams
+          )}. ${sourceInformation}`,
+          program.blockParams.length <= 1
+        );
+
+        let recommendation;
+
+        if (program.blockParams.length === 0) {
+          recommendation = 'Use `{{#if}}` instead.';
+        } else if (inverse) {
+          recommendation = 'Use `{{#let}}` together with `{{#if}} instead.';
+        } else {
+          recommendation =
+            'If you always want the block to render, replace `{{#with}}` with `{{#let}}`. ' +
+            'If you want to conditionally render the block, use `{{#let}}` together with `{{#if}} instead.';
+        }
+
+        deprecate(`\`{{#with}}\` is deprecated. ${recommendation} ${sourceInformation}`, false, {
+          id: 'ember-glimmer.with-syntax',
+          until: '4.0.0',
+          for: 'ember-source',
+          since: {
+            enabled: '3.26.0-beta.1',
+          },
+        });
```

Comment:
We'd need to add a url to the deprecation guide entry. This can probably be done post-merge and I'd like to hand that (writing the deprecation guide) off to someone else so I can focus on implementing other deprecation RFCs.

Reply:
@locks